perceval/backends/_abstract_backends.py

```diff
@@ -70,6 +70,19 @@ def samples(self, count: int):
 
 
 class AProbAmpliBackend(ABackend):
+
+    def __init__(self):
+        super().__init__()
+        self._cache_iterator = dict()
+
+    def _get_iterator(self, input_state: BasicState, mask=None):
+        n_photons = input_state.n
+
+        if n_photons not in self._cache_iterator.keys():
+            self._cache_iterator[n_photons] = list(allstate_iterator(input_state, mask))
```

Comment:
The result of the `allstate_iterator` changes given **n** (number of photons), **m** (number of modes) and the mask.
Indexing the `allstate_iterator` cache only on **n** is doable, only if you invalidate the whole cache (reset it to an empty dict) when **m** (m can change when a new circuit is set) or a mask changes.

Alternatively, you could index on a tuple containing (m, n, mask) - making it less readable maybe.

Reply:
I think I'll rather invalidate the cache If you think it's ok...